File: Week 08/ReverseStringII.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Poplar.Algorithm.WeekSix
{
    /// <summary>
    /// https://leetcode-cn.com/problems/reverse-string-ii/
    /// </summary>
    public class ReverseStringII
    {
        public string ReverseStr(string s, int k)
        {
            char[] a = s.ToCharArray();
            for (int start = 0; start < a.Length; start += 2 * k)
            {
                int i = start, j = Math.Min(start + k - 1, a.Length - 1);
                while (i < j)
                {
                    char tmp = a[i];
                    a[i++] = a[j];
                    a[j--] = tmp;
                }
            }
            return new string(a);
        }
    }
}

File: Week 08/id_011/reverse-string-ii.go
package algorithm00401

func reverseStr(s string, k int) string {
	b := []byte(s)
	for i := 0; i < len(s); i++ {
		j := minInt(i+k, len(s))
		reverse(b[i:j])
	}
	return string(b)
}

func minInt(i, j int) int {
	if i < j {
		return i
	}
	return j
}

func reverse(b []byte) {
	i, j := 0, len(b)-1
	for i < j {
		b[i], b[j] = b[j], b[i]
		i++
		j--
	}
}
